// @quantumblack/kedro-viz/lib/components/ui/toggle/toggle.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _classnames = _interopRequireDefault(require("classnames"));
var _modifiers = _interopRequireDefault(require("../../../utils/modifiers"));
/**
 * Shows a toggle button for code panel
 */
const Toggle = _ref => {
  let {
    id = '',
    checked,
    enabled = true,
    title,
    onChange,
    className = '',
    dataTest
  } = _ref;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _modifiers.default)((0, _classnames.default)('pipeline-toggle', className), {
      enabled
    })
  }, /*#__PURE__*/_react.default.createElement("input", {
    id: "pipeline-toggle-input-".concat(id),
    "data-test": dataTest,
    className: "pipeline-toggle-input",
    type: "checkbox",
    checked: checked,
    disabled: !enabled,
    onChange: onChange
  }), /*#__PURE__*/_react.default.createElement("label", {
    className: (0, _modifiers.default)('pipeline-toggle-label', {
      checked: enabled && checked
    }),
    htmlFor: "pipeline-toggle-input-".concat(id)
  }, title));
};
var _default = exports.default = Toggle;
