// @quantumblack/kedro-viz/lib/components/ui/circle-progress-bar/circle-progress.bar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const FULL_DASH_ARRAY = 283;
const TIME_LIMIT = 60;
const CircleProgressBar = _ref => {
  let {
    children: counter
  } = _ref;
  const calculateTimeFraction = (0, _react.useCallback)(() => {
    return counter / TIME_LIMIT;
  }, [counter]);
  const setCircleDasharray = (0, _react.useCallback)(() => {
    const circleDasharray = "".concat((calculateTimeFraction() * FULL_DASH_ARRAY).toFixed(0), " 283");
    document.getElementById('base-timer-path-remaining').setAttribute('stroke-dasharray', circleDasharray);
  }, [calculateTimeFraction]);
  const startTimer = (0, _react.useCallback)(() => {
    const timerInterval = setInterval(() => {
      setCircleDasharray();
    }, 1000);
    return () => clearInterval(timerInterval);
  }, [setCircleDasharray]);
  (0, _react.useEffect)(() => startTimer(), [startTimer]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "base-timer"
  }, /*#__PURE__*/_react.default.createElement("svg", {
    className: "base-timer__svg",
    viewBox: "0 0 100 100",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/_react.default.createElement("g", {
    className: "base-timer__circle"
  }, /*#__PURE__*/_react.default.createElement("circle", {
    className: "base-timer__path-elapsed",
    cx: "50",
    cy: "50",
    r: "45"
  }), /*#__PURE__*/_react.default.createElement("path", {
    id: "base-timer-path-remaining",
    strokeDasharray: "283",
    className: "base-timer__path-remaining",
    d: " M 50, 50 m -45, 0 a 45,45 0 1,0 90,0 a 45,45 0 1,0 -90,0 "
  }))), /*#__PURE__*/_react.default.createElement("span", {
    id: "base-timer-label",
    className: "base-timer__label"
  }, counter));
};
var _default = exports.default = CircleProgressBar;
