// @quantumblack/kedro-viz/lib/components/export-modal/export-modal.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mapStateToProps = exports.mapDispatchToProps = exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactRedux = require("react-redux");
var _modal = _interopRequireDefault(require("../ui/modal"));
var _button = _interopRequireDefault(require("../ui/button"));
var _actions = require("../../actions");
var _exportGraph = _interopRequireDefault(require("./export-graph"));
/**
 * Modal to allow users to choose between SVG/PNG export formats
 */
const ExportModal = _ref => {
  let {
    graphSize,
    theme,
    onToggle,
    visible
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(_modal.default, {
    closeModal: () => onToggle(false),
    title: "Export pipeline visualisation",
    visible: visible.exportModal
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "pipeline-export-modal"
  }, /*#__PURE__*/_react.default.createElement(_button.default, {
    dataTest: 'export-modal-download-png-btn',
    onClick: () => {
      (0, _exportGraph.default)({
        format: 'png',
        theme,
        graphSize
      });
      onToggle(false);
    }
  }, "Download PNG"), /*#__PURE__*/_react.default.createElement(_button.default, {
    dataTest: 'export-modal-download-svg-btn',
    onClick: () => {
      (0, _exportGraph.default)({
        format: 'svg',
        theme,
        graphSize
      });
      onToggle(false);
    }
  }, "Download SVG")));
};
const mapStateToProps = state => ({
  graphSize: state.graph.size || {},
  visible: state.visible,
  theme: state.theme
});
exports.mapStateToProps = mapStateToProps;
const mapDispatchToProps = dispatch => ({
  onToggle: value => {
    dispatch((0, _actions.toggleExportModal)(value));
  }
});
exports.mapDispatchToProps = mapDispatchToProps;
var _default = exports.default = (0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(ExportModal);
