// @quantumblack/kedro-viz/lib/components/shareable-url-modal/compatibility-error-view/compatibility-error-view.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _button = _interopRequireDefault(require("../../ui/button"));
const CompatibilityErrorView = _ref => {
  let {
    onClick
  } = _ref;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "shareable-url-modal__button-wrapper shareable-url-modal__button-wrapper--right"
  }, /*#__PURE__*/_react.default.createElement(_button.default, {
    mode: "secondary",
    onClick: onClick,
    size: "small"
  }, "Cancel"), /*#__PURE__*/_react.default.createElement("a", {
    href: "https://docs.kedro.org/en/latest/visualisation/share_kedro_viz.html",
    rel: "noreferrer",
    target: "_blank"
  }, /*#__PURE__*/_react.default.createElement(_button.default, {
    size: "small"
  }, "View documentation")));
};
var _default = exports.default = CompatibilityErrorView;
