// @quantumblack/kedro-viz/lib/components/shareable-url-modal/url-box/url-box.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _classnames = _interopRequireDefault(require("classnames"));
var _tooltip = _interopRequireDefault(require("../../ui/tooltip"));
var _button = _interopRequireDefault(require("../../ui/button"));
const UrlBox = _ref => {
  let {
    className,
    url,
    onCopyClick,
    href,
    showCopiedText,
    dataTest
  } = _ref;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)('url-box__wrapper', {
      ["".concat(className)]: className
    })
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "url-box__result-url-wrapper"
  }, /*#__PURE__*/_react.default.createElement("a", {
    className: "url-box__result-url",
    href: href,
    target: "_blank",
    rel: "noopener noreferrer"
  }, url)), window.navigator.clipboard && /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)('url-box___button', {
      'url-box__button-copied': showCopiedText
    })
  }, /*#__PURE__*/_react.default.createElement(_button.default, {
    onClick: () => onCopyClick(url),
    size: "small",
    dataTest: dataTest
  }, "Copy link"), /*#__PURE__*/_react.default.createElement(_tooltip.default, {
    text: "Copied!",
    visible: showCopiedText,
    noDelay: true,
    centerArrow: true,
    arrowSize: "small"
  })));
};
var _default = exports.default = UrlBox;
