// @quantumblack/kedro-viz/lib/actions/tags.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TOGGLE_TAG_FILTER = exports.TOGGLE_TAG_ACTIVE = void 0;
exports.toggleTagActive = toggleTagActive;
exports.toggleTagFilter = toggleTagFilter;
const TOGGLE_TAG_ACTIVE = exports.TOGGLE_TAG_ACTIVE = 'TOGGLE_TAG_ACTIVE';

/**
 * Toggle a tag's highlighting on/off (or array of tags)
 * @param {String|Array} tagIDs Tag id(s)
 * @param {Boolean} active True if tag(s) active
 */
function toggleTagActive(tagIDs, active) {
  return {
    type: TOGGLE_TAG_ACTIVE,
    tagIDs: Array.isArray(tagIDs) ? tagIDs : [tagIDs],
    active
  };
}
const TOGGLE_TAG_FILTER = exports.TOGGLE_TAG_FILTER = 'TOGGLE_TAG_FILTER';

/**
 * Toggle a tag's filtering on/off (or array of tags)
 * @param {String|Array} tagIDs Tag id(s)
 * @param {Boolean} enabled True if tag(s) enabled
 */
function toggleTagFilter(tagIDs, enabled) {
  return {
    type: TOGGLE_TAG_FILTER,
    tagIDs: Array.isArray(tagIDs) ? tagIDs : [tagIDs],
    enabled
  };
}
