// @quantumblack/kedro-viz/lib/reducers/visible.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _actions = require("../actions");
function visibleReducer() {
  let visibleState = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  let action = arguments.length > 1 ? arguments[1] : undefined;
  switch (action.type) {
    case _actions.TOGGLE_GRAPH:
      {
        return Object.assign({}, visibleState, {
          graph: action.visible
        });
      }
    case _actions.TOGGLE_METADATA_MODAL:
      {
        return Object.assign({}, visibleState, {
          metadataModal: action.visible
        });
      }
    case _actions.TOGGLE_EXPORT_MODAL:
      {
        return Object.assign({}, visibleState, {
          exportModal: action.visible
        });
      }
    case _actions.TOGGLE_SHAREABLE_URL_MODAL:
      {
        return Object.assign({}, visibleState, {
          shareableUrlModal: action.visible
        });
      }
    case _actions.TOGGLE_SETTINGS_MODAL:
      {
        return Object.assign({}, visibleState, {
          settingsModal: action.visible
        });
      }
    case _actions.TOGGLE_SIDEBAR:
      {
        return Object.assign({}, visibleState, {
          sidebar: action.visible
        });
      }
    case _actions.TOGGLE_MINIMAP:
      {
        return Object.assign({}, visibleState, {
          miniMap: action.visible
        });
      }
    case _actions.TOGGLE_CODE:
      {
        return Object.assign({}, visibleState, {
          code: action.visible
        });
      }
    case _actions.TOGGLE_MODULAR_PIPELINE_FOCUS_MODE:
      {
        return Object.assign({}, visibleState, {
          modularPipelineFocusMode: action.modularPipeline
        });
      }
    default:
      return visibleState;
  }
}
var _default = exports.default = visibleReducer;
