// @quantumblack/kedro-viz/lib/components/experiment-tracking/accordion/accordion.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactRouterDom = require("react-router-dom");
var _classnames = _interopRequireDefault(require("classnames"));
var _config = require("../../../config");
var _helpers = require("../../../store/helpers");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * A collapsable container component.
 * @param {Object} children React children
 * @param {String|null} className A top-level class name for the component.
 * @param {String} heading Text to display on the top-level.
 * @param {String|null} headingClassName A class name for the accordion header.
 * @param {String|null} headingDetail Text to display on the top-level.
 * @param {Boolean} isCollapsed Control to collapse or expand the content.
 * @param {Boolean} isHyperlink Whether the button has an anchor link to a different page or not
 * @param {String|null} layout A secondary text string for additional context
 * @param {String|null} linkTitle A tag defines the title of the element
 * @param {Function} onCallback Fire a function on click from a parent.
 * @param {String} size Set the header font size.
 */
const Accordion = _ref => {
  let {
    children,
    className = null,
    heading = '',
    headingClassName = null,
    headingDetail = null,
    isCollapsed = false,
    isHyperlink = false,
    layout = 'right',
    linkTitle = null,
    onCallback,
    size = 'small'
  } = _ref;
  const [collapsed, setCollapsed] = (0, _react.useState)(isCollapsed);
  const [flowchartUrl, setFlowchartUrl] = (0, _react.useState)(null);
  const {
    pathname,
    search
  } = (0, _reactRouterDom.useLocation)();
  (0, _react.useEffect)(() => {
    setCollapsed(isCollapsed);
  }, [isCollapsed]);
  (0, _react.useEffect)(() => {
    if (heading.length > 0) {
      const url = (0, _reactRouterDom.generatePath)(_config.routes.flowchart.selectedName, {
        pipelineId: '__default__',
        fullName: heading
      });
      setFlowchartUrl(url);
    }
  }, [heading]);
  const onClick = () => {
    setCollapsed(!collapsed);
    onCallback && onCallback();
  };
  const onLinkToFlowChart = () => {
    (0, _helpers.saveLocalStorage)(_config.localStorageFlowchartLink, {
      fromURL: pathname + search,
      showGoBackBtn: true
    });
  };
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)('accordion', {
      'accordion--left': layout === 'left',
      ["".concat(className)]: className
    })
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)('accordion__heading', {
      ["".concat(headingClassName)]: headingClassName
    })
  }, layout === 'left' && /*#__PURE__*/_react.default.createElement("button", {
    "aria-label": "".concat(collapsed ? 'Show' : 'Hide', " ").concat(heading.toLowerCase()),
    onClick: onClick,
    className: (0, _classnames.default)('accordion__toggle', {
      'accordion__toggle--alt': collapsed
    })
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)('accordion__title', {
      'accordion__title--medium': size === 'medium',
      'accordion__title--large': size === 'large'
    })
  }, isHyperlink ? /*#__PURE__*/_react.default.createElement("a", {
    className: "accordion__title--hyperlink",
    href: flowchartUrl,
    onClick: onLinkToFlowChart,
    title: linkTitle
  }, heading) : heading, headingDetail && /*#__PURE__*/_react.default.createElement("span", {
    className: "accordion__title__detail"
  }, headingDetail)), layout === 'right' && /*#__PURE__*/_react.default.createElement("button", {
    "aria-label": "".concat(collapsed ? 'Show' : 'Hide', " ").concat(heading.toLowerCase()),
    onClick: onClick,
    className: (0, _classnames.default)('accordion__toggle', {
      'accordion__toggle--alt': collapsed
    })
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: collapsed ? 'accordion__content--hide' : null
  }, children));
};
var _default = exports.default = Accordion;
