// @quantumblack/kedro-viz/lib/components/settings-modal/settings-modal-row.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _toggle = _interopRequireDefault(require("../ui/toggle"));
/**
 * Shows a list of MetaDataValue
 */
const SettingsModalRow = _ref => {
  let {
    id,
    name,
    toggleValue,
    description,
    onToggleChange
  } = _ref;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "pipeline-settings-modal__column",
    key: id
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "pipeline-settings-modal__name"
  }, name), /*#__PURE__*/_react.default.createElement(_toggle.default, {
    id: id,
    className: "pipeline-settings-modal__state",
    dataTest: "settings-modal-toggle-".concat(id, "-").concat(toggleValue),
    title: toggleValue ? 'On' : 'Off',
    checked: toggleValue,
    onChange: onToggleChange
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "pipeline-settings-modal__description"
  }, description));
};
var _default = exports.default = SettingsModalRow;
