// @quantumblack/kedro-viz/lib/apollo/config.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.client = void 0;
var _crossFetch = _interopRequireDefault(require("cross-fetch"));
var _client = require("@apollo/client");
var _utilities = require("@apollo/client/utilities");
var _utils = require("../utils");
const httpLink = (0, _client.createHttpLink)({
  // our graphql endpoint, normally here: http://localhost:4141/graphql
  uri: "".concat((0, _utils.sanitizedPathname)(), "graphql"),
  fetch: _crossFetch.default
});
const splitLink = (0, _client.split)(_ref => {
  let {
    query
  } = _ref;
  const definition = (0, _utilities.getMainDefinition)(query);
  return definition.kind === 'OperationDefinition';
}, httpLink);
const client = exports.client = new _client.ApolloClient({
  connectToDevTools: true,
  link: splitLink,
  cache: new _client.InMemoryCache(),
  defaultOptions: {
    query: {
      errorPolicy: 'all'
    },
    mutate: {
      errorPolicy: 'all'
    }
  }
});
