// @quantumblack/kedro-viz/lib/components/experiment-tracking/run-metadata/run-metadata-loader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SingleRunMetadataLoader = exports.MetaDataLoader = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactContentLoader = _interopRequireDefault(require("react-content-loader"));
var _config = require("../../../config");
const GAP = _config.experimentTrackingLazyLoadingGap;
const TitleLoader = () => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("rect", {
  width: "180",
  height: "20",
  x: "0",
  y: "12"
}), /*#__PURE__*/_react.default.createElement("rect", {
  width: "88",
  height: "16",
  x: "0",
  y: 12 + GAP
}), /*#__PURE__*/_react.default.createElement("rect", {
  width: "88",
  height: "16",
  x: "0",
  y: 12 + GAP * 2
}), /*#__PURE__*/_react.default.createElement("rect", {
  width: "88",
  height: "16",
  x: "0",
  y: 12 + GAP * 3
}), /*#__PURE__*/_react.default.createElement("rect", {
  width: "88",
  height: "16",
  x: "0",
  y: 12 + GAP * 4
}), /*#__PURE__*/_react.default.createElement("rect", {
  width: "88",
  height: "16",
  x: "0",
  y: 12 + GAP * 5
}), /*#__PURE__*/_react.default.createElement("rect", {
  width: "88",
  height: "16",
  x: "0",
  y: 12 + GAP * 6
}));
const DetailsLoader = _ref => {
  let {
    x
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("rect", {
    width: "0",
    height: "0",
    x: x,
    y: "12"
  }), /*#__PURE__*/_react.default.createElement("rect", {
    width: "30",
    height: "16",
    x: x,
    y: 12 + GAP
  }), /*#__PURE__*/_react.default.createElement("rect", {
    width: "180",
    height: "16",
    x: x,
    y: 12 + GAP * 2
  }), /*#__PURE__*/_react.default.createElement("rect", {
    width: "88",
    height: "16",
    x: x,
    y: 12 + GAP * 3
  }), /*#__PURE__*/_react.default.createElement("rect", {
    width: "50",
    height: "16",
    x: x,
    y: 12 + GAP * 4
  }), /*#__PURE__*/_react.default.createElement("rect", {
    width: "100",
    height: "16",
    x: x,
    y: 12 + GAP * 5
  }), /*#__PURE__*/_react.default.createElement("rect", {
    width: "150",
    height: "16",
    x: x,
    y: 12 + GAP * 6
  }));
};
const SingleRunMetadataLoader = _ref2 => {
  let {
    theme
  } = _ref2;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "details-metadata"
  }, /*#__PURE__*/_react.default.createElement(_reactContentLoader.default, {
    viewBox: "0 0 1000 300",
    width: "1000px",
    height: "100%",
    backgroundColor: theme === 'dark' ? _config.experimentTrackingLazyLoadingColours.backgroundDarkTheme : _config.experimentTrackingLazyLoadingColours.backgroundLightTheme,
    foregroundColor: theme === 'dark' ? _config.experimentTrackingLazyLoadingColours.foregroundDarkTheme : _config.experimentTrackingLazyLoadingColours.foregroundLightTheme
  }, /*#__PURE__*/_react.default.createElement(TitleLoader, null), /*#__PURE__*/_react.default.createElement(DetailsLoader, {
    x: 380
  })));
};
exports.SingleRunMetadataLoader = SingleRunMetadataLoader;
const MetaDataLoader = _ref3 => {
  let {
    length,
    theme
  } = _ref3;
  const x = length > 1 ? 75 : 0;
  return /*#__PURE__*/_react.default.createElement("tbody", {
    className: "details-metadata__run-lazy-loader"
  }, /*#__PURE__*/_react.default.createElement("tr", null, /*#__PURE__*/_react.default.createElement("td", null, /*#__PURE__*/_react.default.createElement(_reactContentLoader.default, {
    viewBox: "0 0 200 300",
    width: "200px",
    height: "100%",
    backgroundColor: theme === 'dark' ? _config.experimentTrackingLazyLoadingColours.backgroundDarkTheme : _config.experimentTrackingLazyLoadingColours.backgroundLightTheme,
    foregroundColor: theme === 'dark' ? _config.experimentTrackingLazyLoadingColours.foregroundDarkTheme : _config.experimentTrackingLazyLoadingColours.foregroundLightTheme
  }, /*#__PURE__*/_react.default.createElement(DetailsLoader, {
    x: x
  })))));
};
exports.MetaDataLoader = MetaDataLoader;
