// @quantumblack/kedro-viz/lib/components/json-object/json-object.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _modifiers = _interopRequireDefault(require("../../utils/modifiers"));
var _reactJsonView = _interopRequireDefault(require("react-json-view"));
var _jsonTheme = require("./json-theme");
/**
 * Shows a metadata object
 */
const JSONObject = _ref => {
  let {
    className,
    value,
    kind,
    theme,
    empty,
    style,
    collapsed
  } = _ref;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _modifiers.default)('pipeline-json__object', {
      kind
    }, className)
  }, Object.keys(value).length === 0 ? empty : /*#__PURE__*/_react.default.createElement(_reactJsonView.default, {
    theme: theme === 'dark' ? _jsonTheme.darkjsonViewerTheme : _jsonTheme.lightjsonViewerTheme,
    name: false,
    indentWidth: 1,
    collapsed: collapsed ? collapsed : 1,
    collapseStringsAfterLength: true,
    enableClipboard: true,
    displayDataTypes: false,
    src: value,
    style: style ? style : {
      backgroundColor: 'transparent'
    }
  }));
};
var _default = exports.default = JSONObject;
