// @quantumblack/kedro-viz/lib/components/experiment-tracking/run-dataset/run-dataset-loader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SingleRunDatasetLoader = exports.DatasetLoader = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactContentLoader = _interopRequireDefault(require("react-content-loader"));
var _config = require("../../../config");
const GAP = _config.experimentTrackingLazyLoadingGap;
const SubCatLoader = _ref => {
  let {
    y
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("rect", {
    width: "10",
    height: "10",
    x: "0",
    y: y + 3
  }), /*#__PURE__*/_react.default.createElement("rect", {
    width: "100",
    height: "20",
    x: "30",
    y: y
  }));
};
const TitleLoader = _ref2 => {
  let {
    y
  } = _ref2;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("rect", {
    width: "10",
    height: "10",
    x: "0",
    y: y + 3
  }), /*#__PURE__*/_react.default.createElement("rect", {
    width: "180",
    height: "20",
    x: "30",
    y: y
  }), /*#__PURE__*/_react.default.createElement("rect", {
    width: "50",
    height: "16",
    x: "0",
    y: y + GAP
  }), /*#__PURE__*/_react.default.createElement("rect", {
    width: "100",
    height: "16",
    x: "0",
    y: y + GAP * 2
  }));
};
const DetailsLoader = _ref3 => {
  let {
    x,
    y
  } = _ref3;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("rect", {
    width: "0",
    height: "0",
    x: x,
    y: y
  }), /*#__PURE__*/_react.default.createElement("rect", {
    width: "50",
    height: "16",
    x: x,
    y: y + GAP
  }), /*#__PURE__*/_react.default.createElement("rect", {
    width: "180",
    height: "16",
    x: x,
    y: y + GAP * 2
  }));
};
const SingleRunDatasetLoader = _ref4 => {
  let {
    theme
  } = _ref4;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "details-dataset"
  }, /*#__PURE__*/_react.default.createElement(_reactContentLoader.default, {
    viewBox: "0 0 1000 625",
    width: "1000px",
    height: "100%",
    backgroundColor: theme === 'dark' ? _config.experimentTrackingLazyLoadingColours.backgroundDarkTheme : _config.experimentTrackingLazyLoadingColours.backgroundLightTheme,
    foregroundColor: theme === 'dark' ? _config.experimentTrackingLazyLoadingColours.foregroundDarkTheme : _config.experimentTrackingLazyLoadingColours.foregroundLightTheme
  }, /*#__PURE__*/_react.default.createElement(SubCatLoader, {
    y: 0
  }), /*#__PURE__*/_react.default.createElement(TitleLoader, {
    y: 50
  }), /*#__PURE__*/_react.default.createElement(DetailsLoader, {
    x: 380,
    y: 50
  }), /*#__PURE__*/_react.default.createElement(TitleLoader, {
    y: 185
  }), /*#__PURE__*/_react.default.createElement(DetailsLoader, {
    x: 380,
    y: 185
  }), /*#__PURE__*/_react.default.createElement(TitleLoader, {
    y: 320
  }), /*#__PURE__*/_react.default.createElement(DetailsLoader, {
    x: 380,
    y: 320
  })));
};
exports.SingleRunDatasetLoader = SingleRunDatasetLoader;
const DatasetLoader = _ref5 => {
  let {
    x,
    y,
    length,
    theme
  } = _ref5;
  return /*#__PURE__*/_react.default.createElement(_reactContentLoader.default, {
    viewBox: "0 10 200 30",
    width: "180px",
    height: "100%",
    backgroundColor: theme === 'dark' ? _config.experimentTrackingLazyLoadingColours.backgroundDarkTheme : _config.experimentTrackingLazyLoadingColours.backgroundLightTheme,
    foregroundColor: theme === 'dark' ? _config.experimentTrackingLazyLoadingColours.foregroundDarkTheme : _config.experimentTrackingLazyLoadingColours.foregroundLightTheme
  }, /*#__PURE__*/_react.default.createElement("rect", {
    width: "150",
    height: "16",
    x: x,
    y: y + length * 2
  }));
};
exports.DatasetLoader = DatasetLoader;
