// @quantumblack/kedro-viz/lib/components/ui/button/button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _classnames = _interopRequireDefault(require("classnames"));
/**
 * Generic Kedro Button
 */
const Button = _ref => {
  let {
    children,
    dataTest = 'test-default-btn',
    disabled = false,
    onClick,
    size = 'regular',
    mode = 'primary'
  } = _ref;
  return /*#__PURE__*/_react.default.createElement("span", {
    className: "kedro button"
  }, /*#__PURE__*/_react.default.createElement("button", {
    className: (0, _classnames.default)('button__btn', "button__btn--".concat(size), "button__btn--".concat(mode)),
    "data-test": dataTest,
    disabled: disabled,
    onClick: onClick
  }, children));
};
Button.propTypes = {
  dataTest: _propTypes.default.string,
  disabled: _propTypes.default.bool,
  mode: _propTypes.default.oneOf(['primary', 'secondary', 'success']),
  onClick: _propTypes.default.func,
  size: _propTypes.default.oneOf(['regular', 'small'])
};
var _default = exports.default = Button;
