// @quantumblack/kedro-viz/lib/components/experiment-tracking/run-plots-modal/run-plots-modal.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _plotlyChart = _interopRequireDefault(require("../../plotly-chart"));
var _backWide = _interopRequireDefault(require("../../icons/back-wide"));
var _nodeIcon = _interopRequireDefault(require("../../icons/node-icon"));
var _shortType = _interopRequireDefault(require("../../../utils/short-type"));
var _dateUtils = require("../../../utils/date-utils");
var _classnames = _interopRequireDefault(require("classnames"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const RunPlotsModal = _ref => {
  let {
    runDatasetToShow,
    visible,
    setShowRunPlotsModal
  } = _ref;
  const {
    datasetKey,
    datasetType,
    datasetValues
  } = runDatasetToShow;
  const runDataWithPlotData = datasetValues === null || datasetValues === void 0 ? void 0 : datasetValues.filter(_ref2 => {
    let {
      value
    } = _ref2;
    return value;
  });
  const numDatasets = runDataWithPlotData === null || runDataWithPlotData === void 0 ? void 0 : runDataWithPlotData.length;
  const plotView = numDatasets === 3 ? 'threeCharts' : numDatasets === 2 ? 'twoCharts' : 'oneChart';
  const isPlotly = (0, _shortType.default)(datasetType) === 'plotly';
  const isImage = (0, _shortType.default)(datasetType) === 'image';
  const nodeTypeIcon = (0, _shortType.default)(datasetType);
  const handleKeyDown = event => {
    if (event.keyCode === 27) {
      setShowRunPlotsModal(false);
    }
  };
  (0, _react.useEffect)(() => {
    if (visible) {
      window.addEventListener('keydown', handleKeyDown);
    }
    return () => window.removeEventListener('keydown', handleKeyDown);
  });
  if (!visible) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "pipeline-run-plots-modal"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "pipeline-run-plots-modal__top"
  }, /*#__PURE__*/_react.default.createElement("button", {
    className: "pipeline-run-plots-modal__back",
    onClick: () => setShowRunPlotsModal(false)
  }, /*#__PURE__*/_react.default.createElement(_backWide.default, {
    className: "pipeline-run-plots-modal__back-icon"
  }), /*#__PURE__*/_react.default.createElement("span", {
    className: "pipeline-run-plots-modal__back-text"
  }, "Back")), /*#__PURE__*/_react.default.createElement("div", {
    className: "pipeline-run-plots-modal__header"
  }, /*#__PURE__*/_react.default.createElement(_nodeIcon.default, {
    className: "pipeline-run-plots-modal__icon",
    icon: nodeTypeIcon
  }), /*#__PURE__*/_react.default.createElement("span", {
    className: "pipeline-run-plots-modal__title"
  }, datasetKey))), /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)('pipeline-run-plots-modal__content', "pipeline-run-plots-modal__content--".concat(plotView), {
      'pipeline-run-plots-modal__content--plotly': isPlotly
    })
  }, isPlotly && runDataWithPlotData.map(data => {
    return data.value && /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _classnames.default)('pipeline-run-plots__plot-wrapper', "pipeline-run-plots__plot-wrapper--".concat(plotView)),
      key: data.runId
    }, /*#__PURE__*/_react.default.createElement(_plotlyChart.default, {
      data: data.value.data,
      layout: data.value.layout,
      view: plotView
    }), /*#__PURE__*/_react.default.createElement("div", {
      className: "pipeline-run-plots-image-text"
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: "pipeline-run-plots-image-text-timestamp"
    }, data.runId), /*#__PURE__*/_react.default.createElement("span", {
      className: "pipeline-run-plots-image-text-relative_timestamp"
    }, (0, _dateUtils.toHumanReadableTime)(data.runId))));
  }), isImage && runDataWithPlotData.map(data => {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _classnames.default)("pipeline-run-plots__image-wrapper", "pipeline-run-plots__image-wrapper--".concat(plotView)),
      key: data.runId
    }, /*#__PURE__*/_react.default.createElement("img", {
      alt: "Matplotlib rendering",
      className: (0, _classnames.default)("pipeline-run-plots__image--".concat(plotView)),
      src: "data:image/png;base64,".concat(data.value)
    }), /*#__PURE__*/_react.default.createElement("div", {
      className: "pipeline-run-plots-image-text"
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: "pipeline-run-plots-image-text-timestamp"
    }, data.runId), /*#__PURE__*/_react.default.createElement("span", {
      className: "pipeline-run-plots-image-text-relative_timestamp"
    }, (0, _dateUtils.toHumanReadableTime)(data.runId))));
  })));
};
var _default = exports.default = RunPlotsModal;
