// @quantumblack/kedro-viz/lib/components/node-list/node-list-row-list.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _modifiers = _interopRequireDefault(require("../../utils/modifiers"));
var _nodeListRow = _interopRequireWildcard(require("./node-list-row"));
var _lazyList = _interopRequireDefault(require("../lazy-list"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const NodeRowList = _ref => {
  let {
    items = [],
    group,
    collapsed,
    onItemClick,
    onItemChange,
    onItemMouseEnter,
    onItemMouseLeave
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(_lazyList.default, {
    height: (start, end) => (end - start) * _nodeListRow.nodeListRowHeight,
    total: items.length
  }, _ref2 => {
    let {
      start,
      end,
      total,
      listRef,
      upperRef,
      lowerRef,
      listStyle,
      upperStyle,
      lowerStyle
    } = _ref2;
    return /*#__PURE__*/_react.default.createElement("ul", {
      ref: listRef,
      style: listStyle,
      className: (0, _modifiers.default)('pipeline-nodelist__children', {
        closed: collapsed
      }, 'pipeline-nodelist__list pipeline-nodelist__list--nested')
    }, /*#__PURE__*/_react.default.createElement("li", {
      className: (0, _modifiers.default)('pipeline-nodelist__placeholder-upper', {
        fade: start !== end && start > 0
      }),
      ref: upperRef,
      style: upperStyle
    }), /*#__PURE__*/_react.default.createElement("li", {
      className: (0, _modifiers.default)('pipeline-nodelist__placeholder-lower', {
        fade: start !== end && end < total
      }),
      ref: lowerRef,
      style: lowerStyle
    }), items.slice(start, end).map(item => /*#__PURE__*/_react.default.createElement(_nodeListRow.default, {
      container: "li",
      key: item.id,
      id: item.id,
      kind: group.kind,
      label: item.highlightedLabel,
      count: item.count,
      name: item.name,
      type: item.type,
      icon: item.icon,
      active: item.active,
      checked: item.checked,
      disabled: item.disabled,
      faded: item.faded,
      visible: item.visible,
      selected: item.selected,
      allUnchecked: group.allUnchecked,
      visibleIcon: item.visibleIcon,
      invisibleIcon: item.invisibleIcon,
      onClick: () => onItemClick(item),
      onMouseEnter: () => onItemMouseEnter(item),
      onMouseLeave: () => onItemMouseLeave(item),
      onChange: e => onItemChange(item, !e.target.checked),
      rowType: "filter"
    })));
  });
};
var _default = exports.default = NodeRowList;
