// @quantumblack/kedro-viz/lib/components/shareable-url-modal/main-view/main-view.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _classnames = _interopRequireDefault(require("classnames"));
var _dropdown = _interopRequireDefault(require("../../ui/dropdown"));
var _button = _interopRequireDefault(require("../../ui/button"));
var _iconButton = _interopRequireDefault(require("../../ui/icon-button"));
var _info = _interopRequireDefault(require("../../icons/info"));
var _input = _interopRequireDefault(require("../../ui/input"));
var _menuOption = _interopRequireDefault(require("../../ui/menu-option"));
var _toggle = _interopRequireDefault(require("../../ui/toggle"));
var _config = require("../../../config");
/* eslint-disable camelcase */

const renderTextContent = (isPreviewEnabled, setIsPreviewEnabled) => {
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "shareable-url-modal__content-wrapper"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "shareable-url-modal__content-title"
  }, "Publish and Share Kedro-Viz"), /*#__PURE__*/_react.default.createElement("h2", {
    className: "shareable-url-modal__content-description-title"
  }, "Prerequisites", ' '), /*#__PURE__*/_react.default.createElement("p", {
    className: "shareable-url-modal__content-description"
  }, "Deploying and hosting Kedro-Viz requires access keys or user credentials, depending on the chosen service provider. To use this feature, please add your access keys or credentials as environment variables in your project. More information can be found in the", ' ', /*#__PURE__*/_react.default.createElement("a", {
    target: "_blank",
    rel: "noopener noreferrer",
    href: _config.KEDRO_VIZ_PUBLISH_DOCS_URL
  }, "documentation"), "."), /*#__PURE__*/_react.default.createElement("h2", {
    className: "shareable-url-modal__content-description-title shareable-url-modal__content-description-title--disclaimer"
  }, "Disclaimer", ' '), /*#__PURE__*/_react.default.createElement("p", {
    className: "shareable-url-modal__content-description"
  }, "Disclaimer Kedro-Viz contains preview data for multiple datasets. You can enable or disable all previews when publishing Kedro-Viz."), /*#__PURE__*/_react.default.createElement("div", {
    className: "shareable-url-modal__content-preview-dataset"
  }, "All dataset previews", /*#__PURE__*/_react.default.createElement(_toggle.default, {
    className: "shareable-url-modal__content-toggle",
    dataTest: "shareable-url-modal-preview-dataset-".concat(isPreviewEnabled),
    title: isPreviewEnabled ? 'On' : 'Off',
    checked: isPreviewEnabled,
    onChange: () => setIsPreviewEnabled(prev => !prev)
  })));
};
const MainView = _ref => {
  let {
    handleModalClose,
    handleSubmit,
    inputValues,
    isFormDirty,
    onPlatformChange,
    onBuckNameChange,
    onEndpointChange,
    setIsPreviewEnabled,
    isPreviewEnabled,
    visible
  } = _ref;
  const {
    platform,
    bucket_name,
    endpoint
  } = inputValues || {};
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
    className: "shareable-url-modal__content-form-wrapper"
  }, renderTextContent(isPreviewEnabled, setIsPreviewEnabled), /*#__PURE__*/_react.default.createElement("div", {
    className: "shareable-url-modal__form-wrapper"
  }, /*#__PURE__*/_react.default.createElement("p", {
    className: "shareable-url-modal__form-wrapper-title"
  }, "Please enter the required information below."), /*#__PURE__*/_react.default.createElement("div", {
    className: "shareable-url-modal__input-wrapper"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "shareable-url-modal__input-label"
  }, "Hosting platform"), /*#__PURE__*/_react.default.createElement(_dropdown.default, {
    defaultText: platform && _config.hostingPlatforms[platform],
    placeholderText: !platform ? 'Select a hosting platform' : null,
    onChanged: onPlatformChange,
    dataTest: 'shareable-url-modal-dropdown-hosting-platform',
    width: null
  }, Object.entries(_config.hostingPlatforms).map(_ref2 => {
    let [value, label] = _ref2;
    return /*#__PURE__*/_react.default.createElement(_menuOption.default, {
      className: (0, _classnames.default)({
        'pipeline-list__option--active': platform === value
      }),
      key: value,
      primaryText: label,
      value: value
    });
  }))), /*#__PURE__*/_react.default.createElement("div", {
    className: "shareable-url-modal__input-wrapper"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "shareable-url-modal__input-label"
  }, "Bucket name"), /*#__PURE__*/_react.default.createElement(_input.default, {
    defaultValue: bucket_name,
    onChange: onBuckNameChange,
    placeholder: "Enter name",
    resetValueTrigger: visible,
    size: "small",
    type: "input",
    dataTest: 'shareable-url-modal-input-bucket-name'
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: "shareable-url-modal__input-wrapper"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "shareable-url-modal__endpoint-url-wrapper"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "shareable-url-modal__input-label"
  }, "Endpoint URL"), /*#__PURE__*/_react.default.createElement(_iconButton.default, {
    ariaLabel: "The endpoint URL information",
    className: "shareable-url-modal__information-icon",
    labelText: /*#__PURE__*/_react.default.createElement("p", {
      className: "shareable-url-modal__information-text"
    }, "The endpoint URL is the link to where your Kedro-Viz will be hosted. For information on obtaining the endpoint URL, please refer to the documentation for", ' ', /*#__PURE__*/_react.default.createElement("a", {
      className: "shareable-url-modal__input-label-text",
      href: _config.KEDRO_VIZ_PUBLISH_AWS_DOCS_URL,
      rel: "noreferrer",
      target: "_blank"
    }, "AWS"), ",", ' ', /*#__PURE__*/_react.default.createElement("a", {
      className: "shareable-url-modal__input-label-text",
      href: _config.KEDRO_VIZ_PUBLISH_AZURE_DOCS_URL,
      rel: "noreferrer",
      target: "_blank"
    }, "Azure"), ",", ' ', /*#__PURE__*/_react.default.createElement("a", {
      className: "shareable-url-modal__input-label-text",
      href: _config.KEDRO_VIZ_PUBLISH_GCP_DOCS_URL,
      rel: "noreferrer",
      target: "_blank"
    }, "GCP")),
    icon: _info.default
  })), /*#__PURE__*/_react.default.createElement(_input.default, {
    defaultValue: endpoint,
    onChange: onEndpointChange,
    placeholder: "Enter url",
    resetValueTrigger: visible,
    size: "small",
    type: "input",
    dataTest: 'shareable-url-modal-input-endpoint'
  })))), /*#__PURE__*/_react.default.createElement("div", {
    className: "shareable-url-modal__button-wrapper"
  }, /*#__PURE__*/_react.default.createElement(_button.default, {
    mode: "secondary",
    onClick: handleModalClose,
    size: "small",
    dataTest: 'shareable-url-modal-cancel-btn'
  }, "Cancel"), /*#__PURE__*/_react.default.createElement(_button.default, {
    disabled: !Object.values(isFormDirty).every(value => value),
    size: "small",
    onClick: handleSubmit,
    dataTest: 'shareable-url-modal-publish-publish-btn'
  }, "Publish")));
};
var _default = exports.default = MainView;
