// @quantumblack/kedro-viz/lib/components/experiment-tracking/experiment-primary-toolbar/sliding-animation.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SlideFromRightToLeft = exports.SlideFromLeftToRight = exports.Animation = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactTransitionGroup = require("react-transition-group");
const directions = {
  leftToRight: {
    entering: {
      transform: 'translateX(0)',
      visibility: 'visible',
      opacity: '1'
    },
    entered: {
      transform: 'translateX(0)',
      visibility: 'visible',
      opacity: '1'
    },
    exiting: {
      transform: 'translateX(-34%)',
      visibility: 'hidden',
      opacity: '0'
    },
    exited: {
      transform: 'translateX(-34%)',
      visibility: 'hidden',
      opacity: '0'
    }
  },
  rightToLeft: {
    entering: {
      transform: 'translateX(0)',
      visibility: 'visible',
      opacity: '1'
    },
    entered: {
      transform: 'translateX(0)',
      visibility: 'visible',
      opacity: '1'
    },
    exiting: {
      transform: 'translateX(34%)',
      visibility: 'hidden',
      opacity: '0'
    },
    exited: {
      transform: 'translateX(34%)',
      visibility: 'hidden',
      opacity: '0'
    }
  }
};
const Animation = _ref => {
  let {
    children,
    direction,
    duration,
    state
  } = _ref;
  const defaultStyle = {
    transition: "transform 0.3s ease-out, opacity 0.1s linear"
  };
  return /*#__PURE__*/_react.default.createElement(_reactTransitionGroup.Transition, {
    in: state,
    timeout: duration
  }, state => /*#__PURE__*/_react.default.createElement("div", {
    style: {
      ...defaultStyle,
      ...directions[direction][state]
    }
  }, children));
};
exports.Animation = Animation;
const SlideFromLeftToRight = _ref2 => {
  let {
    state,
    duration,
    children
  } = _ref2;
  return /*#__PURE__*/_react.default.createElement(Animation, {
    children: children,
    direction: "leftToRight",
    duration: duration,
    state: state
  });
};
exports.SlideFromLeftToRight = SlideFromLeftToRight;
const SlideFromRightToLeft = _ref3 => {
  let {
    state,
    duration,
    children
  } = _ref3;
  return /*#__PURE__*/_react.default.createElement(Animation, {
    children: children,
    direction: "rightToLeft",
    duration: duration,
    state: state
  });
};
exports.SlideFromRightToLeft = SlideFromRightToLeft;
