// @quantumblack/kedro-viz/lib/components/experiment-tracking/tooltip/tooltip.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.tooltipDefaultProps = exports.ExperimentTrackingTooltip = void 0;
var _react = _interopRequireDefault(require("react"));
var _classnames = _interopRequireDefault(require("classnames"));
const tooltipDefaultProps = exports.tooltipDefaultProps = {
  content: {
    label1: '',
    value1: '',
    label2: '',
    value2: ''
  },
  direction: 'right',
  position: {
    x: -500,
    y: -500
  },
  visible: false
};
const ExperimentTrackingTooltip = _ref => {
  let {
    content = tooltipDefaultProps.content,
    direction = tooltipDefaultProps.direction,
    position = tooltipDefaultProps.position,
    visible = tooltipDefaultProps.visible
  } = _ref;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)('tooltip', {
      'tooltip--show': visible
    }),
    style: {
      transform: "translate(".concat(position.x, "px, ").concat(position.y, "px)")
    }
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: (0, _classnames.default)('tooltip-arrow', "tooltip-arrow--".concat(direction))
  }), /*#__PURE__*/_react.default.createElement("h3", {
    className: "tooltip-label"
  }, "".concat(content === null || content === void 0 ? void 0 : content.label1, ":")), /*#__PURE__*/_react.default.createElement("h4", {
    className: "tooltip-value"
  }, content === null || content === void 0 ? void 0 : content.value1), /*#__PURE__*/_react.default.createElement("br", null), /*#__PURE__*/_react.default.createElement("h3", {
    className: "tooltip-label"
  }, "".concat(content === null || content === void 0 ? void 0 : content.label2, ":")), /*#__PURE__*/_react.default.createElement("h4", {
    className: "tooltip-value"
  }, content === null || content === void 0 ? void 0 : content.value2));
};
exports.ExperimentTrackingTooltip = ExperimentTrackingTooltip;
