// @quantumblack/kedro-viz/lib/components/metadata/metadata-url.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));

/**
 * Shows metadata label and a clickable value (hyperlink) with improved styling
 */
const MetaDataURL = _ref => {
  let {
    label,
    href,
    value,
    theme,
    visible = true,
    empty = '-',
    title = null
  } = _ref;

  return visible && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("dt", {
    className: "pipeline-metadata__label"
  }, label), /*#__PURE__*/_react.default.createElement("dd", {
    className: "pipeline-metadata__row",
    "data-label": label,
    style: { marginBottom: '10px' }
  }, href && value ? /*#__PURE__*/_react.default.createElement("a", {
    href: href,
    target: "_blank",
    rel: "noopener noreferrer",
    title: title || value,
    className: `pipeline-metadata__link ${theme || ''}`,
    style: {
      color: '#0056b3', // Brighter blue for readability
      textDecoration: 'none',
      backgroundColor: '#f0f8ff', // Light background for contrast
      padding: '5px 10px',
      borderRadius: '4px',
      fontSize: '14px',
      display: 'inline-block'
    },
    onMouseOver: e => (e.target.style.backgroundColor = '#e0f0ff'), // Lighter hover effect
    onMouseOut: e => (e.target.style.backgroundColor = '#f0f8ff')
  }, value) : /*#__PURE__*/_react.default.createElement("span", {
    style: { color: '#6c757d', fontStyle: 'italic' } // Subtle style for empty state
  }, empty)));
};

var _default = exports.default = MetaDataURL;
