// @quantumblack/kedro-viz/lib/components/metadata/metadata-row.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _metadataList = _interopRequireDefault(require("./metadata-list"));
var _jsonObject = _interopRequireDefault(require("../json-object"));
var _metadataValue = _interopRequireDefault(require("./metadata-value"));
/**
 * Shows metadata label and a given single value, or a list of values, or child elements
 */
const MetaDataRow = _ref => {
  let {
    label,
    theme,
    property,
    value,
    kind = 'text',
    empty = '-',
    visible = true,
    inline = true,
    commas = true,
    limit = false,
    title = null,
    children
  } = _ref;
  const showList = Array.isArray(value);
  const showObject = typeof value === 'object' && value !== null && !showList;
  return visible && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("dt", {
    className: "pipeline-metadata__label"
  }, label), /*#__PURE__*/_react.default.createElement("dd", {
    className: "pipeline-metadata__row",
    "data-label": label
  }, showList && /*#__PURE__*/_react.default.createElement(_metadataList.default, {
    property: property,
    inline: inline,
    commas: commas,
    kind: kind,
    empty: empty,
    values: value,
    limit: limit
  }), !showList && !showObject && !children && /*#__PURE__*/_react.default.createElement(_metadataValue.default, {
    empty: empty,
    kind: kind,
    theme: theme,
    title: title || value,
    value: value
  }), showObject && /*#__PURE__*/_react.default.createElement(_jsonObject.default, {
    value: value,
    kind: kind,
    theme: theme,
    empty: empty
  }), children));
};
var _default = exports.default = MetaDataRow;
